

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>API Walkthrough &mdash; loompy 0.30.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="loompy 0.30.0 documentation" href="../index.html"/>
        <link rel="next" title="Complete API Refrence" href="../fullapi/index.html"/>
        <link rel="prev" title="Understanding the semantics of loom files" href="../semantics/index.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> loompy
          

          
          </a>

          
            
            
              <div class="version">
                0.30
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to loompy!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../semantics/index.html">Understanding the semantics of loom files</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Walkthrough</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#creating-and-connecting">Creating and connecting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#creating-loom-files">Creating <code class="docutils literal"><span class="pre">.loom</span></code> files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#connecting-to-loom-files">Connecting to <code class="docutils literal"><span class="pre">.loom</span></code> files</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#manipulate-data">Manipulate data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#shape-indexing-and-slicing">Shape, indexing and slicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#global-attributes">Global attributes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#row-and-column-attributes">Row and column attributes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-attributes-and-columns">Adding attributes and columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#operations">Operations</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#map">Map</a></li>
<li class="toctree-l4"><a class="reference internal" href="#permutation">Permutation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#batch-scan">Batch scan</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#layers">Layers</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#working-with-layers">Working with layers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#create-a-layer">Create a layer</a></li>
<li class="toctree-l4"><a class="reference internal" href="#access-a-layer">Access a layer</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fullapi/index.html">Complete API Refrence</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../format/index.html"><code class="docutils literal"><span class="pre">.loom</span></code> file format</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">loompy</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>API Walkthrough</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/apiwalkthrough/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="api-walkthrough">
<span id="apiwalkthrough"></span><h1>API Walkthrough<a class="headerlink" href="#api-walkthrough" title="Permalink to this headline">¶</a></h1>
<div class="section" id="creating-and-connecting">
<span id="loomcreate"></span><h2>Creating and connecting<a class="headerlink" href="#creating-and-connecting" title="Permalink to this headline">¶</a></h2>
<div class="section" id="creating-loom-files">
<h3>Creating <code class="docutils literal"><span class="pre">.loom</span></code> files<a class="headerlink" href="#creating-loom-files" title="Permalink to this headline">¶</a></h3>
<p>Create from data:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">matrix</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">row_attrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> <span class="n">col_attrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> <span class="n">file_attrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chunks</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">64</span><span class="p">,</span> <span class="mi">64</span><span class="p">),</span> <span class="n">chunk_cache</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">512</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">,</span> <span class="n">compression_opts</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">LoomConnection</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a new .loom file from the given data.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str):         The filename (typically using a `.loom` file extension)</span>
<span class="sd">        matrix (numpy.ndarray): Two-dimensional (N-by-M) numpy ndarray of float values</span>
<span class="sd">        row_attrs (dict):       Row attributes, where keys are attribute names and values</span>
<span class="sd">                                are numpy arrays (float or string) of length N</span>
<span class="sd">        col_attrs (dict):       Column attributes, where keys are attribute names and</span>
<span class="sd">                                values are numpy arrays (float or string) of length M</span>
<span class="sd">        chunks (tuple):         The chunking of the matrix. Small chunks are slow</span>
<span class="sd">                                when loading a large batch of rows/columns in sequence,</span>
<span class="sd">                                but fast for single column/row retrieval.</span>
<span class="sd">                                Defaults to (64,64).</span>
<span class="sd">        chunk_cache (int):      Sets the chunk cache used by the HDF5 format inside</span>
<span class="sd">                                the loom file, in MB. If the cache is too small to</span>
<span class="sd">                                contain all chunks of a row/column in memory, then</span>
<span class="sd">                                sequential row/column access will be a lot slower.</span>
<span class="sd">                                Defaults to 512.</span>
<span class="sd">        dtype (str):           Dtype of the matrix. Default float32 (uint16, float16 could be used)</span>
<span class="sd">        compression_opts (int): Strenght of the gzip compression. Default None.</span>
<span class="sd">    Returns:</span>
<span class="sd">        LoomConnection to created loom file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>Create by combining existing .loom files:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">combine</span><span class="p">(</span><span class="n">files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">file_attrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Combine two or more loom files and save as a new loom file</span>

<span class="sd">    Args:</span>
<span class="sd">        files (list of str):    the list of input files (full paths)</span>
<span class="sd">        output_file (str):      full path of the output loom file</span>
<span class="sd">        key (string):           Row attribute to use to verify row ordering</span>
<span class="sd">        file_attrs (dict):      file attributes (title, description, url, etc.)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Nothing, but creates a new loom file combining the input files.</span>

<span class="sd">    The input files must (1) have exactly the same number of rows, (2) have</span>
<span class="sd">    exactly the same sets of row and column attributes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>Create from a 10X Genomics
<a class="reference external" href="http://support.10xgenomics.com/single-cell/software/pipelines/latest/what-is-cell-ranger">cellranger</a>
output folder:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_from_cellranger</span><span class="p">(</span><span class="n">folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">loom_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cell_id_prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">sample_annotation</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;mm10&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">LoomConnection</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a .loom file from 10X Genomics cellranger output</span>

<span class="sd">    Args:</span>
<span class="sd">        folder (str):               path to the cellranger output folder (usually called `outs`)</span>
<span class="sd">        loom_file (str):            full path of the resulting loom file</span>
<span class="sd">        cell_id_prefix (str):       prefix to add to cell IDs (e.g. the sample id for this sample)</span>
<span class="sd">        sample_annotation (dict):   dict of additional sample attributes</span>
<span class="sd">        genome (str):               genome build to load (e.g. &#39;mm10&#39;)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Nothing, but creates loom_file</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>You can use the <em>sample_annotation</em> dictionary to add column (cell)
annotations to all cells in the dataset. For example, this is useful to
add a sample ID to each of several datasets before combining them into a
single .loom file.</p>
</div>
<div class="section" id="connecting-to-loom-files">
<h3>Connecting to <code class="docutils literal"><span class="pre">.loom</span></code> files<a class="headerlink" href="#connecting-to-loom-files" title="Permalink to this headline">¶</a></h3>
<p>Establish a connection to an existing <code class="docutils literal"><span class="pre">.loom</span></code> file:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">connect</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;r+&#39;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">LoomConnection</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Establish a connection to a .loom file.</span>

<span class="sd">    Args:</span>
<span class="sd">        filename (str):     Name of the .loom file to open</span>
<span class="sd">        mode (str):         read/write mode, accepts &#39;r+&#39; (read/write) or</span>
<span class="sd">                            &#39;r&#39; (read-only), defaults to &#39;r+&#39;</span>

<span class="sd">    Returns:</span>
<span class="sd">        A LoomConnection instance.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>Example:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">loompy</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="s2">&quot;filename.loom&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>In the rest of the documentation below, <code class="docutils literal"><span class="pre">ds</span></code> is assumed to be an
instance of <code class="docutils literal"><span class="pre">LoomConnection</span></code> obtained by connecting to a <code class="docutils literal"><span class="pre">.loom</span></code>
file.</p>
<p>Note: there is usually no need to close the connection. The exception is
if you need to write to the loom file from two different processes
(sequentially, not simultaneously). In that case, the first process
needs to let go of the file by calling <code class="docutils literal"><span class="pre">close()</span></code> on the connection,
before the second can start writing:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="manipulate-data">
<span id="loommanipulate"></span><h2>Manipulate data<a class="headerlink" href="#manipulate-data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="shape-indexing-and-slicing">
<h3>Shape, indexing and slicing<a class="headerlink" href="#shape-indexing-and-slicing" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">shape</span></code> property returns the row and column count as a tuple:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(100,2345)</span>
</pre></div>
</div>
<p>The data stored in the main matrix can be retrieved by indexing and
slicing. The following are supported:</p>
<ul class="simple">
<li>Indices: anything that can be converted to a Python long</li>
<li>Slices (i.e. <code class="docutils literal"><span class="pre">:</span></code> or <code class="docutils literal"><span class="pre">0:10</span></code>)</li>
<li>Lists of the rows/columns you want (i.e. <code class="docutils literal"><span class="pre">[0,</span> <span class="pre">34,</span> <span class="pre">576]</span></code>)</li>
<li>Mask arrays (i.e. numpy array of bool indicating the rows/columns you
want)</li>
</ul>
<p>Lists and mask arrays are supported along one dimension at a time only.
Note that performance will be poor if you select many rows (columns) out
of a large matrix. It may be better to load the entire matrix and then
perform the sub-selection in memory (using numpy slicing).</p>
<p>Since the main matrix is two-dimensional, two arguments are always
needed. Examples:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="p">[:,</span> <span class="p">:]</span>          <span class="c1"># Return the entire matrix</span>
<span class="n">ds</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>    <span class="c1"># Return the 10x10 submatrix starting at row and column zero</span>
<span class="n">ds</span><span class="p">[</span><span class="mi">99</span><span class="p">,</span> <span class="p">:]</span>         <span class="c1"># Return the 100th row</span>
<span class="n">ds</span><span class="p">[:,</span> <span class="mi">99</span><span class="p">]</span>         <span class="c1"># Return the 100th column</span>
<span class="n">ds</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="p">:]</span>    <span class="c1"># Return rows with index 0, 3 and 5</span>
<span class="n">ds</span><span class="p">[:,</span> <span class="n">bool_array</span><span class="p">]</span> <span class="c1"># Return columns where bool_array elements are True</span>
</pre></div>
</div>
</div>
<div class="section" id="global-attributes">
<h3>Global attributes<a class="headerlink" href="#global-attributes" title="Permalink to this headline">¶</a></h3>
<p>Global attributes are available as</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;title&quot;</span><span class="p">]</span>
<span class="go">&quot;The title of the dataset&quot;</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;title&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;New title&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;title&quot;</span><span class="p">]</span>
<span class="go">&quot;New title&quot;</span>
</pre></div>
</div>
<p>The following global attributes are standard:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">title</span></code>, a short title for the dataset</li>
<li><code class="docutils literal"><span class="pre">description</span></code>, a longer description of the dataset</li>
<li><code class="docutils literal"><span class="pre">url</span></code>, a link to a web page for the dataset</li>
<li><code class="docutils literal"><span class="pre">doi</span></code>, a DOI for the paper where the dataset was published</li>
</ul>
<p>(They are standard in the sense that you are encouraged to use <code class="docutils literal"><span class="pre">title</span></code>
rather than <code class="docutils literal"><span class="pre">Title</span></code> or <code class="docutils literal"><span class="pre">TITLE</span></code> for a title, but they are not
guaranteed to exist, or required)</p>
<p>The following global attributes are reserved:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">schema</span></code>, a type annotation schema (JSON-formatted string)</li>
</ul>
<p>DO NOT attempt to set reserved global attributes to a different value.</p>
</div>
<div class="section" id="row-and-column-attributes">
<h3>Row and column attributes<a class="headerlink" href="#row-and-column-attributes" title="Permalink to this headline">¶</a></h3>
<p>Row and column attributes are accessed as dictionaries on <code class="docutils literal"><span class="pre">row_attrs</span></code>
and <code class="docutils literal"><span class="pre">col_attrs</span></code>, respectively. For example:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">row_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>       <span class="c1"># Return list of row attribute names</span>
<span class="n">ds</span><span class="o">.</span><span class="n">col_attrs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>       <span class="c1"># Return list of column attribute names</span>
<span class="n">ds</span><span class="o">.</span><span class="n">row_attrs</span><span class="p">[</span><span class="s2">&quot;GeneName&quot;</span><span class="p">]</span>  <span class="c1"># Return a numpy array of gene names (assuming the attribute exists)</span>
</pre></div>
</div>
<p>Note that these dictionaries are <strong>read-only</strong>. Any modifications will
not be saved in the .loom file and will cause internal inconsistencies
in the <code class="docutils literal"><span class="pre">LoomConnection</span></code> object. Use <em>set_attr()</em> (below) to add or
modify attributes.</p>
<p>For convenience, attributes are also available directly on the
<code class="docutils literal"><span class="pre">LoomConnection</span></code> object:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">GeneName</span>     <span class="c1"># Equivalent to ds.row_attrs[&quot;GeneName&quot;]</span>
</pre></div>
</div>
<p>Using attributes in this way results in a very compact and readable
syntax for selecting subarrays:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="p">[</span><span class="n">ds</span><span class="o">.</span><span class="n">Gene</span> <span class="o">==</span> <span class="s2">&quot;Actb&quot;</span><span class="p">,:]</span>
<span class="go">array([[  2.,   9.,   9., ...,   0.,  14.,   0.]], dtype=float32)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">logical_or</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">Gene</span> <span class="o">==</span> <span class="s2">&quot;Actb&quot;</span><span class="p">,</span> <span class="n">ds</span><span class="o">.</span><span class="n">Gene</span> <span class="o">==</span> <span class="s2">&quot;Gapdh&quot;</span><span class="p">),:]</span>
<span class="go">array([[  2.,   9.,   9., ...,   0.,  14.,   0.],</span>
<span class="go">       [  0.,   1.,   4., ...,   0.,  14.,   3.]], dtype=float32)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="p">[:,</span> <span class="n">ds</span><span class="o">.</span><span class="n">CellID</span> <span class="o">==</span> <span class="s2">&quot;AAACATACATTCTC-1&quot;</span><span class="p">]</span>
<span class="go">array([[ 0.],</span>
<span class="go">       [ 0.],</span>
<span class="go">       [ 0.],</span>
<span class="go">       ...,</span>
<span class="go">       [ 0.],</span>
<span class="go">       [ 0.],</span>
<span class="go">       [ 0.]], dtype=float32)</span>
</pre></div>
</div>
<p>There are some limitations:</p>
<ul class="simple">
<li>Custom attributes do not override existing <code class="docutils literal"><span class="pre">LoomConnection</span></code>
attributes, such as method names. For example, if your .loom file has
a row attribute <code class="docutils literal"><span class="pre">shape</span></code>, then <code class="docutils literal"><span class="pre">ds.shape</span></code> will not return that
attribute, but will still return the shape of the main matrix.</li>
<li>Column attributes take precedence. For example, if you have both
<code class="docutils literal"><span class="pre">ds.row_attrs[&quot;Name&quot;]</span></code> and <code class="docutils literal"><span class="pre">ds.col_attrs[&quot;Name&quot;]</span></code>, then
<code class="docutils literal"><span class="pre">ds.Name</span></code> returns the column attribute, not the row attribute.</li>
</ul>
<p>Note again, that you should not assign to these attributes, because your
assignment will not be saved in the .loom file and will cause internal
inconsistencies in the <code class="docutils literal"><span class="pre">LoomConnection</span></code> object. Use <em>set_attr()</em>
(below) to add or modify attributes.</p>
</div>
<div class="section" id="adding-attributes-and-columns">
<h3>Adding attributes and columns<a class="headerlink" href="#adding-attributes-and-columns" title="Permalink to this headline">¶</a></h3>
<p>You can add attributes and columns to an existing loom file. It is not
possible to add rows or to delete attributes or any part of the matrix.</p>
<p>To add an attribute, which also saves it to the loom file:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">set_attr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create or modify an attribute.</span>

<span class="sd">    Args:</span>
<span class="sd">        name (str):             Name of the attribute</span>
<span class="sd">        values (numpy.ndarray): Array of values of length equal to the axis length</span>
<span class="sd">        axis (int):             Axis of the attribute (0 = rows, 1 = columns)</span>
<span class="sd">        dtype (str):            Type (&quot;float64&quot;, &quot;int&quot;, or &quot;string&quot;)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Nothing.</span>

<span class="sd">    This will overwrite any existing attribute of the same name.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p><strong>Note:</strong> If you use an existing attribute name, the existing attribute
will be overwritten. This is pefectly fine, and is the only way to
change an attribute or its type.</p>
<p>To add columns:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">add_columns</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">submatrix</span><span class="p">,</span> <span class="n">col_attrs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add columns of data and attribute values to the dataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        submatrix (numpy.ndarray):  An N-by-M matrix of floats (N rows, M columns)</span>
<span class="sd">        col_attrs (dict):           Column attributes, where keys are attribute names and values are numpy arrays (float or string) of length M</span>

<span class="sd">    Returns:</span>
<span class="sd">        Nothing.</span>

<span class="sd">    Note that this will modify the underlying HDF5 file, which will interfere with any concurrent readers.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>You need to provide a submatrix corresponding to the columns, as well as
a dictionary of column attributes with values for all the new columns.</p>
<p><strong>Note:</strong> It is not possible to add rows.</p>
<p>You can also add the contents of another .loom file:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">add_loom</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">key</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">fill_values</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add the content of another loom file</span>

<span class="sd">    Args:</span>
<span class="sd">        other_file (str):   filename of the loom file to append</span>
<span class="sd">        fill_values (dict): default values to use for missing attributes (or None to drop missing attrs, or &#39;auto&#39; to fill with sensible defaults)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Nothing, but adds the loom file. Note that the other loom file must have exactly the same</span>
<span class="sd">        number of rows, and must have exactly the same column attributes.</span>
<span class="sd">        The all the contents including layers but ignores layers in `other_file` that are not already persent in self</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>The content of the other file is added as columns on the right of the
current dataset. The rows must match for this to work. That is, the two
files must have exactly the same rows (genes). If <code class="docutils literal"><span class="pre">key</span></code> is given, the
rows may be out of order, and will be aligned based on the key
attribute. Furthermore, the two datasets must have the same column
attributes (but of coure can have different <em>values</em> for those
attributes at each column). Missing attributes can be given default
values using <code class="docutils literal"><span class="pre">fill_values</span></code> .</p>
</div>
<div class="section" id="operations">
<span id="loomoperations"></span><h3>Operations<a class="headerlink" href="#operations" title="Permalink to this headline">¶</a></h3>
<div class="section" id="map">
<h4>Map<a class="headerlink" href="#map" title="Permalink to this headline">¶</a></h4>
<p>You can map a function across all rows (all columns), while avoiding
loading the entire dataset into memory:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">map</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">f_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Callable</span><span class="p">[[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> <span class="nb">int</span><span class="p">]],</span> <span class="n">axis</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">chunksize</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">selection</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Apply a function along an axis without loading the entire dataset in memory.</span>

<span class="sd">    Args:</span>
<span class="sd">        f (list of func):       Function(s) that takes a numpy ndarray as argument</span>

<span class="sd">        axis (int):     Axis along which to apply the function (0 = rows, 1 = columns)</span>

<span class="sd">        chunksize (int): Number of rows (columns) to load per chunk</span>

<span class="sd">        selection (array of bool): Columns (rows) to include</span>

<span class="sd">    Returns:</span>
<span class="sd">        numpy.ndarray result of function application</span>

<span class="sd">        If you supply a list of functions, the result will be a list of numpy arrays. This is more</span>
<span class="sd">        efficient than repeatedly calling map() one function at a time.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
<p>The function will receive an array (of floats) as its only argument, and
should return a single float value.</p>
<p>Example:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ds</span><span class="o">.</span><span class="n">map</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
<span class="go"># Returns an array of row means</span>
<span class="go">np.array([1.23, 0.32, ...])</span>
</pre></div>
</div>
</div>
<div class="section" id="permutation">
<h4>Permutation<a class="headerlink" href="#permutation" title="Permalink to this headline">¶</a></h4>
<p>Permute the order of the rows (or columns):</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">permute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ordering</span><span class="p">,</span> <span class="n">axis</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Permute the dataset along the indicated axis.</span>

<span class="sd">    Args:</span>
<span class="sd">        ordering (list of int):     The desired order along the axis</span>
<span class="sd">        axis (int):                 The axis along which to permute</span>

<span class="sd">    Returns:</span>
<span class="sd">        Nothing.</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="batch-scan">
<h4>Batch scan<a class="headerlink" href="#batch-scan" title="Permalink to this headline">¶</a></h4>
<p>For very large loom files, it’s very useful to scan across the file
(along either rows or columns) in <em>batches</em>, to avoid loading the entire
file in memory. This can be achieved using the <code class="docutils literal"><span class="pre">batch_scan</span></code> method:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">batch_scan</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cells</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">axis</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;Performs a batch scan of the loom file</span>

<span class="sd">    Args</span>
<span class="sd">    ----</span>
<span class="sd">    cells: np.ndarray</span>
<span class="sd">        the indexes [1,2,3,..,1000] of the cells to select</span>
<span class="sd">    genes: np.ndarray</span>
<span class="sd">        the indexes [1,2,3,..,1000] of the genes to select</span>
<span class="sd">    axis: int</span>
<span class="sd">        0:rows or 1:cols</span>
<span class="sd">    batch_size: int</span>
<span class="sd">        the chuncks returned at every element of the iterator</span>

<span class="sd">    Returns</span>
<span class="sd">    ------</span>
<span class="sd">    Iterable that yields triplets</span>
<span class="sd">    (ix, indexes, vals)</span>

<span class="sd">    ix: int</span>
<span class="sd">        first position / how many rows/cols have been yielded alredy</span>
<span class="sd">    indexes: np.ndarray[int]</span>
<span class="sd">        the indexes with the same numbering of the input args cells / genes (i.e. np.arange(len(ds.shape[axis])))</span>
<span class="sd">        this is ix + selection</span>
<span class="sd">    vals: np.ndarray</span>
<span class="sd">        the matrix corresponding to the chunk</span>
<span class="sd">    &quot;&quot;&quot;</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="layers">
<span id="loomlayers"></span><h2>Layers<a class="headerlink" href="#layers" title="Permalink to this headline">¶</a></h2>
<div class="section" id="working-with-layers">
<h3>Working with layers<a class="headerlink" href="#working-with-layers" title="Permalink to this headline">¶</a></h3>
<p>Loom supports multiple layers. There is always a single main matrix, but
optionally one or more additional layers having the same number of rows
and columns. Layers are accessed using the <code class="docutils literal"><span class="pre">layer</span></code> property on the
<code class="docutils literal"><span class="pre">LoomConnection</span></code>.</p>
<div class="section" id="create-a-layer">
<h4>Create a layer<a class="headerlink" href="#create-a-layer" title="Permalink to this headline">¶</a></h4>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">set_layer</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">matrix</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">chunks</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">64</span><span class="p">,</span> <span class="mi">64</span><span class="p">),</span> <span class="n">chunk_cache</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">512</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;float32&quot;</span><span class="p">,</span> <span class="n">compression_opts</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</pre></div>
</div>
</div>
<div class="section" id="access-a-layer">
<h4>Access a layer<a class="headerlink" href="#access-a-layer" title="Permalink to this headline">¶</a></h4>
<p>The <code class="docutils literal"><span class="pre">layer</span></code> property returns a Layer object, which can be sliced to
get the data:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">layer</span><span class="p">[</span><span class="s2">&quot;layer&quot;</span><span class="p">][</span><span class="mi">10</span><span class="p">,</span> <span class="p">:]</span>
</pre></div>
</div>
<p>The default layer can be accessed directly:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="p">:]</span>
</pre></div>
</div>
<p>It can also be accessed using the empty string:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">layer</span><span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Layers can be loaded in memory as sparse matrices, efficiently:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">LoomLayer</span><span class="o">.</span><span class="n">as_coo</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">sparse</span><span class="o">.</span><span class="n">coo_matrix</span><span class="p">:</span>
<span class="n">LoomLayer</span><span class="o">.</span><span class="n">as_csr</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">:</span>
<span class="n">LoomLayer</span><span class="o">.</span><span class="n">as_csc</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">sparse</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">:</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../fullapi/index.html" class="btn btn-neutral float-right" title="Complete API Refrence" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../semantics/index.html" class="btn btn-neutral" title="Understanding the semantics of loom files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, LinnarssonLab.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'0.30.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>