

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>.loom file format &mdash; loompy 0.30.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="loompy 0.30.0 documentation" href="../index.html"/>
        <link rel="prev" title="loompy package" href="../fullapi/loompy.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> loompy
          

          
          </a>

          
            
            
              <div class="version">
                0.30
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to loompy!</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation/index.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../semantics/index.html">Understanding the semantics of loom files</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../apiwalkthrough/index.html">API Walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="../fullapi/index.html">Complete API Refrence</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#"><code class="docutils literal"><span class="pre">.loom</span></code> file format</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#specification">Specification</a></li>
<li class="toctree-l2"><a class="reference internal" href="#hdf5-concepts">HDF5 concepts</a></li>
<li class="toctree-l2"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">loompy</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li><code class="docutils literal"><span class="pre">.loom</span></code> file format</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/format/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="loom-file-format">
<span id="format"></span><h1><code class="docutils literal"><span class="pre">.loom</span></code> file format<a class="headerlink" href="#loom-file-format" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<span id="formatinfo"></span><h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">.loom</span></code> file format is designed to efficiently hold large omics
datasets. Typically, such data takes the form of a large matrix of
numbers, along with metadata for the rows and columns. For example,
single-cell RNA-seq data consists of expression measurements for all
genes (rows) in a large number of cells (columns), along with metadata
for genes (e.g. <code class="docutils literal"><span class="pre">Chromosome</span></code>, <code class="docutils literal"><span class="pre">Strand</span></code>, <code class="docutils literal"><span class="pre">Location</span></code>, <code class="docutils literal"><span class="pre">Name</span></code>), and
for cells (e.g. <code class="docutils literal"><span class="pre">Species</span></code>, <code class="docutils literal"><span class="pre">Sex</span></code>, <code class="docutils literal"><span class="pre">Strain</span></code>, <code class="docutils literal"><span class="pre">GFP</span> <span class="pre">positive</span></code>).</p>
<p>We designed <code class="docutils literal"><span class="pre">.loom</span></code> files to represent such datasets in a way that
treats rows and columns the same. You may want to cluster both genes and
cells, you may want to perform PCA on both of them, and filter based on
quality controls. SQL databases and other data storage solutions almost
always treat data as a <em>table</em>, not a matrix, and makes it very hard to
add arbitrary metadata to rows and columns. In contrast, <code class="docutils literal"><span class="pre">.loom</span></code> makes
this very easy.</p>
<p>Furthermore, current and future datasets can have tens of thousands of
rows (genes) and hundreds of thousands of columns (cells). We designed
<code class="docutils literal"><span class="pre">.loom</span></code> for efficient access to arbitrary rows and columns.</p>
<p>The annotated matrix format lends itself to very natural representation
of common analysis tasks. For example, the result of a clustering
algorithm can be stored simply as another attribute that gives the
cluster ID for each cell. Dimensionality reduction such as PCA or t-SNE,
similarly, can be stored as two attributes giving the projection
coordinates of each cell.</p>
<p>Finally, we recognize the importance of graph-based analyses of such
datasets. Loom supports graphs of both the rows (e.g. genes) and the
columns (e.g. cells), and multiple graphs can be stored each file.</p>
</div>
<div class="section" id="specification">
<span id="specifications"></span><h2>Specification<a class="headerlink" href="#specification" title="Permalink to this headline">¶</a></h2>
<p>A valid <code class="docutils literal"><span class="pre">.loom</span></code> file conforms to the following:</p>
<ul class="simple">
<li>There MUST be a single dataset at <code class="docutils literal"><span class="pre">/matrix</span></code></li>
<li>There can OPTIONALLY be a subgroup <code class="docutils literal"><span class="pre">/layers</span></code> containing additional
matrices (called “layers”)</li>
<li>Each additional layer MUST have the same (N, M)&nbsp;shape</li>
<li>Each layer can have a different data type, compression, chunking etc.</li>
<li>There can OPTIONALLY be at least one <a class="reference external" href="https://www.hdfgroup.org/HDF5/Tutor/crtatt.html">HDF5
attribute</a> on the
root <code class="docutils literal"><span class="pre">/</span></code> group, which MUST be of type <code class="docutils literal"><span class="pre">string</span></code> and should be
interpreted as attributes of the whole <code class="docutils literal"><span class="pre">.loom</span></code> file. The following
HDF5 attributes are standard:</li>
<li><code class="docutils literal"><span class="pre">title</span></code>, a short title for the dataset</li>
<li><code class="docutils literal"><span class="pre">description</span></code>, a longer description of the dataset</li>
<li><code class="docutils literal"><span class="pre">url</span></code>, a link to a web page for the dataset</li>
<li><code class="docutils literal"><span class="pre">doi</span></code>, a DOI for the paper where the dataset was published</li>
<li>There MUST be a group <code class="docutils literal"><span class="pre">/row_attrs</span></code></li>
<li>There can OPTIONALLY be one or more datasets at <code class="docutils literal"><span class="pre">/row_attrs/{name}</span></code>
of length N and type <code class="docutils literal"><span class="pre">float64</span></code>, <code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">string</span></code></li>
<li>There MUST be a group <code class="docutils literal"><span class="pre">/col_attrs</span></code></li>
<li>There can OPTIONALLY be one or more datasets at <code class="docutils literal"><span class="pre">/col_attrs/{name}</span></code>
of length M and type <code class="docutils literal"><span class="pre">float64</span></code>, <code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">string</span></code></li>
</ul>
<p>The datasets under <code class="docutils literal"><span class="pre">/row_attrs</span></code> should be semantically interpreted as
row attributes, with one value per row of the main matrix, and in the
same order. Therefore, all datasets under this group must be
one-dimensional arrays with exactly N elements, where N is the number of
rows in the main matrix.</p>
<p>The datasets under <code class="docutils literal"><span class="pre">/col_attrs</span></code> should be semantically interpreted as
column attributes, with one value per column of the main matrix, and in
the same order. Therefore, all datasets under this group must be
one-dimensional arrays with exactly M elements, where M is the number of
columns in the main matrix.</p>
<p>As noted above, only three datatypes are allowedfor attributes;
<code class="docutils literal"><span class="pre">float64</span></code>, <code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">string</span></code>.</p>
</div>
<div class="section" id="hdf5-concepts">
<span id="hd5concepts"></span><h2>HDF5 concepts<a class="headerlink" href="#hdf5-concepts" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">.loom</span></code> format is based on
<a class="reference external" href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">HDF5</a>, a
standard for storing large numerical datasets. Quoting from h5py.org:</p>
<blockquote>
<div>An HDF5 file is a container for two kinds of objects: datasets,
which are array-like collections of data, and groups, which are
folder-like containers that hold datasets and other groups. The most
fundamental thing to remember when using h5py is: <em>Groups work like
dictionaries, and datasets work like NumPy arrays</em>.</div></blockquote>
<p>A valid <code class="docutils literal"><span class="pre">.loom</span></code> file is simply an HDF5 file that contains specific
<em>groups</em> representing the main matrix as well as row and column
attributes. Because of this, <code class="docutils literal"><span class="pre">.loom</span></code> files can be created and read by
any language that supports HDF5, including <a class="reference external" href="http://h5py.org">Python</a>,
<a class="reference external" href="http://bioconductor.org/packages/release/bioc/html/rhdf5.html">R</a>,
<a class="reference external" href="http://se.mathworks.com/help/matlab/low-level-functions.html">MATLAB</a>,
<a class="reference external" href="https://reference.wolfram.com/language/ref/format/HDF5.html">Mathematica</a>,
<a class="reference external" href="https://www.hdfgroup.org/HDF5/doc/index.html">C</a>,
<a class="reference external" href="https://www.hdfgroup.org/HDF5/doc/cpplus_RM/">C++</a>,
<a class="reference external" href="https://www.hdfgroup.org/products/java/">Java</a>, and
<a class="reference external" href="https://rubygems.org/gems/hdf5/versions/0.3.5">Ruby</a>.</p>
</div>
<div class="section" id="example">
<span id="loomexample"></span><h2>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h2>
<p>Here’s an example of the structure of a valid <code class="docutils literal"><span class="pre">.loom</span></code> file:</p>
<table border="1" class="docutils">
<colgroup>
<col width="22%" />
<col width="32%" />
<col width="46%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Group</th>
<th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>/matrix</td>
<td>float32[N,M] or uint16[N,M]</td>
<td>Main matrix of N rows and M columns</td>
</tr>
<tr class="row-odd"><td>/layers/</td>
<td>(subgroup)</td>
<td>Subgroup of additional matrix layers</td>
</tr>
<tr class="row-even"><td>/row_attrs/</td>
<td>(subgroup)</td>
<td>Subgroup of all row attributes</td>
</tr>
<tr class="row-odd"><td>/row_attrs/Name</td>
<td>string[N]</td>
<td>Row attribute “Name” of type string</td>
</tr>
<tr class="row-even"><td>/col_attrs/</td>
<td>(subgroup)</td>
<td>Subgroup of all column attributes</td>
</tr>
<tr class="row-odd"><td>/col_attrs/CellID</td>
<td>float64[M]</td>
<td>Column attribute “CellID” of type float64</td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../fullapi/loompy.html" class="btn btn-neutral" title="loompy package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, LinnarssonLab.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'0.30.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>