Example directory structure:
```
base_dir
├── data
│   ├── group1
│   │   ├── sample_sheet.yaml
│   │   ├── data1.fastq
│   │   └── data2.fastq
│   └── group1
│       ├── sample_sheet.yaml
│       ├── data1.fastq
│       └── data2.fastq
└── targets
    └── locus_name
        ├── manifest.yaml
        ├── refs.fasta
        ├── refs.fasta.fai
        └── refs.gff
```
Input fastqs and sample sheets describing the experiments that produced them are kept in base_dir/data in 'group' directories.

Each sample_sheet.yaml should contain {experiment names: {description dictionary}} pairs.
Each experiment's description dictionary must always contain the key-value pair:

    target_info: name of the target (see below)

For a PacBio experiment, it should also contain

    fastq_fn: name of the fastq file (relative to the group's directory, i.e. just the basename)

For an illumina experiment, it should contain:
    R1_fn: name of the R1 fastq file (relative to the group's directory, i.e. just the basename)
    R2_fn: name of the R2 fastq file (relative to the group's directory, i.e. just the basename)
    
It can also contain other optional keys describing properties of the experiment, e.g.:
    cell_line: string
    donor_type: string
    replicate: integer
    sorted: boolean

target_info should be the name of a directory in base_dir/targets (e.g. locus_name above) containing descriptions of the sequences used in the experiment.
Nomenclature: 'target' is the genomic locus being targeted, and 'donor' is the exogenous sequence being added.

The required files are:

- a fasta file `refs.fasta` containing the target sequence, donor sequence, and any suspected contaminants

- an index `refs.fasta.fai` of this fasta file

- a gff file `refs.gff` marking various features on these sequences.

    Required features are -
    target: [sgRNA, 5' HA, 3' HA, forward primer, reverse primer]
    donor: [5' HA, 3' HA] (TODO: this is incomplete)

- `manifest.yaml`, with contents

    target: seqname in refs.fasta of target
    donor: seqname in refs.fasta of donor
    sources: list of genbank files from which to generate refs.\*

The fasta, fai, and gff  files can be generated by annotating the target and donor sequences with the required features in benchling, downloading the benchling files in genbank format into the locus_name directory, listing these files in the 'sources' entry in 'manifest.yaml', then running TargetInfo.make_references (TODO: explain).
